aeneas/src/ir/PackingSolver.v3

```diff
@@ -0,0 +1,339 @@
+		var intervals = Vector<Interval>.new();
+
+		for (i < size - s + 1) {
+			var start = size - s - i;
+			var isEmpty = true;
+			for (j < s) {
+				if (bits[start + j] != Bit.Unassigned) {
+					isEmpty = false;
+					break;
+				}
+			}
+			if (isEmpty) intervals.put(Interval(start, start + s));
+		}
+		return intervals.extract();
+	}
+	def assignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unusable;
+	}
+	def tryAssignInterval(i: Interval) -> bool {
+		for (j = i.s; j < i.e; j++) if (bits[j] != Bit.Unassigned) return false;
+		assignInterval(i);
+		return true;
+	}
+	def unassignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unassigned;
+	}
+	def copy() -> PackingPattern { return PackingPattern.new(Arrays.dup(bits)); }
+}
+
+type PackingField {
+	case Nonref(size: byte);
+	case Ref;
+}
+
+type CaseField(caseIdx: int, fieldIdx: int) {
+	def hash() -> int { return caseIdx * 31 + fieldIdx; }
+}
+
+class PackingSolver(size: byte, refPatterns: RefPatterns) {
+	var cases: Array<Array<PackingField>>;
+	var fieldOrder: Array<CaseField>;
+	var assignments = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+
+	def reset() {
+		cases = null;
+		fieldOrder = null;
+		assignments = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+	}
+
+	def canDistinguish(state: Array<PackingPattern>, elements: Array<bool>) -> bool {
+		// TODO: split on bits with the most unassigned
+		var numElements = 0;
+		for (i in elements) if (i) numElements++;
+		if (numElements <= 1) return true;
+
+		for (i < size) {
+			// try to split on this bit
+			var numZeros = 0, numOnes = 0, numUnassigned = 0;
+			for (j < state.length) {
+				if (!elements[j]) continue;
+				match (state[j].bits[i]) {
+					Fixed(v) => if (v == 0) numZeros++; else numOnes++;
+					Unassigned => numUnassigned++;
+					Unusable => ;
+				}
+			}
+
+			if (numUnassigned == 0 && (numZeros == 0 || numOnes == 0)) continue; // useless
+			if (numUnassigned == 1 && numZeros == 0 && numOnes == 0) continue;
+
+			var zeros = Array<bool>.new(elements.length);
+			var ones = Array<bool>.new(elements.length);
+			var parity: u1 = 0;
+			var temporaryBits = Vector<int>.new();
+
+			for (j < state.length) {
+				if (!elements[j]) continue;
+				match (state[j].bits[i]) {
+					Unusable => zeros[j] = ones[j] = true;
+					Fixed(v) => {
+						if (v == 0) zeros[j] = true;
+						else ones[j] = true;
+					}
+					Unassigned => {
+						temporaryBits.put(j);
+						// XXX: do more than alternate the unassigned bits in this column
+						state[j].bits[i] = Bit.Fixed(parity);
+						if (parity == 0) zeros[j] = true;
+						else ones[j] = true;
+						parity++;
+					}
+				}
+			}
+
+			if (canDistinguish(state, zeros) && canDistinguish(state, ones)) return true;
+			for (j < temporaryBits.length) state[temporaryBits[j]].bits[i] = Bit.Unassigned;
+		}
+		return false;
+	}
+
+	def checkDistinguishable(state: Array<PackingPattern>) -> bool {
```

Comment:
This has side-effects, so maybe `trySimpleTaggingHeuristic` and then split off the other heuristic into another method?